Improve operators

nalimilan opened the pull request:
Introduce a new `null_safe_op()` function which is used instead of
`isbits()` to decide whether to take the fast path (without a branch)
or the slow path in lifted operators. This is more correct (and thus
safer), and allows supporting operations on any type. Also add an
extensive set of tests for both paths.

On Julia 0.4, operators for which no functor exist cannot be identified
by `null_safe_op()`: for these, fall back to the slow path on that version.

This PR introduces  `cbrt` as a new supported operation since it has a special
operator `∛`. It also keeps `abs()`, `abs2()`, `scalarmin()` and `scalarmax()`, which
should probably be deprecated once we have a more general solution for
lifting.

Tests pass on Julia 0.4, but on 0.5 they need https://github.com/JuliaLang/julia/pull/16995 (unrelated tests already fail too).

Fixes https://github.com/JuliaStats/NullableArrays.jl/issues/74, https://github.com/JuliaStats/NullableArrays.jl/issues/111, https://github.com/JuliaStats/NullableArrays.jl/issues/116.

nalimilan committed:
Improve operators

Introduce a new null_safe_op() function which is used instead of
isbits() to decide whether to take the fast path (without a branch)
or the slow path in lifted operators. This is more correct (and thus
safer), and allows supporting operations on any type. Also add an
extensive set of tests for both paths.

On Julia 0.4, operators for which no functor exist cannot be identified
by null_safe_op(): for these, fall back to the slow path on that version.

This PR introduces cbrt as a new supported operation since it has a special
operator ∛. It also keeps abs(), abs2(), scalarmin() and scalarmax(), which
should probably be deprecated once we have a more general solution for
lifting.

johnmyleswhite commented on src/operators.jl:
```diff
@@ -1,121 +1,130 @@
+    null_safe_op(f::Any, ::Type)::Bool
+    null_safe_op(f::Any, ::Type, ::Type)::Bool
+
+Returns whether an operation `f` can safely be applied to any value of the passed type(s).
```
Minor nit: maybe we should distinguish between value and bit pattern? Consider the following type:

```
immutable OneType
    x::Int

    function OneType()
        new(1)
    end
end
```

The only legal value of this type is `1`, but our uninitialized constructors will generate bit patterns that are not legal values.

nalimilan replied:
Maybe we should mention "bit pattern" below instead of "domain"? I think "value" is good as a first approximation to give the idea of the requirements.

nalimilan replied:
OK, I've updated the docstring.

nalimilan committed:
Support Nullable{Union{}} with operators

Nullable{Union{}} is always null so we can return null immediately
if one of the arguments is of that type. For unary operators, return
Nullabl(). For binary operators, return Nullable{S}() with S the element
type of the other argument, which carries more information.

nalimilan commented:
With support merged in Julia, tests now pass. Will tag a new release.
